### PR TITLE
[2.0] Add new methods Collection::findFirst and Collection::reduce

### PR DESCRIPTION
**Pull request by VincentLanglet**
This is two methods I find useful for the next major.

Collection::find has a better performance than using Collection::filter then Collection::first.

Collection::reduce is similar to array_reduce.

I create this PR to open the discussion about these methods.

### CODE REVIEW

File: lib/Doctrine/Common/Collections/Collection.php
Comment (greg0ire) [inline lib/Doctrine/Common/Collections/Collection.php]:
```suggestion
     * Returns the first element of this collection that satisfies the predicate p.
```

File: lib/Doctrine/Common/Collections/Collection.php
Comment (greg0ire) [inline lib/Doctrine/Common/Collections/Collection.php]:
```suggestion
     * @return mixed The first element respecting the predicate,
```

File: tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php
Comment (greg0ire) [inline tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php]:
```suggestion
        }), 'Element does not exist');
```

**Review by greg0ire — COMMENTED** [review]
You should probably add docs at https://github.com/doctrine/collections/blob/1.6.x/docs/en/index.rst

**Comment by SenseException** [comment]
> You should probably add docs at https://github.com/doctrine/collections/blob/1.6.x/docs/en/index.rst

Probably the 2.0 docs if this is for the next major release.

**Comment by VincentLanglet** [comment]
I added the documentation.

File: tests/Doctrine/Tests/Common/Collections/BaseCollectionTest.php
Comment (SenseException) [inline tests/Doctrine/Tests/Common/Collections/BaseCollectionTest.php:82]: Can you expand this with more elements please? It should make the test more expressive.

File: tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php
Comment (SenseException) [inline tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php]: Please split up into separate tests for "found" and "not found".

File: tests/Doctrine/Tests/Common/Collections/BaseCollectionTest.php
Comment (SenseException) [inline tests/Doctrine/Tests/Common/Collections/BaseCollectionTest.php]: Please split up into separate tests for "found" and "not found".

File: docs/en/index.rst
Comment (SenseException) [inline docs/en/index.rst]: Can you please use an example where the element, that is searched for, is twice or more in the collection? It would emphasize the chosen behavior.

**Review by SenseException — CHANGES_REQUESTED** [review]

File: tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php
Comment (VincentLanglet) [reply tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php]: All the others tests are not split.
- `testRemove` try to remove 5 elements
- `testEmpty` is called on empty and not empty
- `testExist` is almost a copy paste of my test

And so on...

I feel like it would be inconsistent to split this one.

File: docs/en/index.rst
Comment (VincentLanglet) [reply docs/en/index.rst]: Sure, done

File: tests/Doctrine/Tests/Common/Collections/BaseCollectionTest.php
Comment (VincentLanglet) [reply tests/Doctrine/Tests/Common/Collections/BaseCollectionTest.php:82]: Sure, done.

File: tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php
Comment (SenseException) [reply tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php]: Should have been, but fixing those is out of scope for your PR.

Comment (VincentLanglet) [reply tests/Doctrine/Tests/Common/Collections/BaseArrayCollectionTest.php]: Done

File: docs/en/index.rst
Comment (greg0ire) [inline docs/en/index.rst]: Maybe we should name it `findOne` (as opposed to `findAll()`?)

File: docs/en/index.rst
Comment (greg0ire) [inline docs/en/index.rst]: It would be great to have an example which uses `$key`

File: docs/en/index.rst
Comment (greg0ire) [inline docs/en/index.rst]: Let's use type declarations?

File: docs/en/index.rst
Comment (greg0ire) [inline docs/en/index.rst]: Type declarations here too please?

File: lib/Doctrine/Common/Collections/Collection.php
Comment (greg0ire) [inline lib/Doctrine/Common/Collections/Collection.php:223]: Where can I learn more about this `=` notation? It's the first time I see it I think

File: lib/Doctrine/Common/Collections/Collection.php
Comment (greg0ire) [inline lib/Doctrine/Common/Collections/Collection.php:220]: What if the first element respecting the predicate is literally `false` itself? Wouldn't it be better to throw an exception that recommends to call whatever method allows to check that there will be a result? (`exists`?)

File: lib/Doctrine/Common/Collections/Collection.php
Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php:223]: I don't know. I don't use psalm.
Every other phpdoc use this notation so I did the same.

Comment (greg0ire) [reply lib/Doctrine/Common/Collections/Collection.php:223]: Apparently it makes them optional: 744afc199626c1638113da4cd4bf2222c91dcd13

File: lib/Doctrine/Common/Collections/Collection.php
Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php:220]: I don't like to require an `exists` call, it add complexity to the code and it lower performance (the search is made twice).

Collection is not made to store boolean.
All the method `current/next/first/last/...` return `T|false`.

But maybe, It's better to return `null`.
The method `get` is returning `T|null`.

Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php:220]: I changed for `null` if not found.

File: docs/en/index.rst
Comment (VincentLanglet) [reply docs/en/index.rst]: Done

File: lib/Doctrine/Common/Collections/Collection.php
Comment (greg0ire) [reply lib/Doctrine/Common/Collections/Collection.php:220]: >  All the method current/next/first/last/... return T|false.

Ok, if this is an issue, it's an issue of doctrine/collections, not this PR

**Review by greg0ire — DISMISSED** [review]

File: lib/Doctrine/Common/Collections/Collection.php
Comment (someniatko) [reply lib/Doctrine/Common/Collections/Collection.php:223]: @greg0ire https://psalm.dev/docs/annotating_code/type_syntax/callable_types/

**Review by greg0ire — DISMISSED** [review]

**Review by SenseException — DISMISSED** [review]

File: lib/Doctrine/Common/Collections/Collection.php
Comment (ostrolucky) [inline lib/Doctrine/Common/Collections/Collection.php]: I'm wondering how was new `findOne` method name chosen. Name doesn't indicate it will return first matching result. For all I know, it can return last matching result. In fact, it's doing same thing as `first()`, except it skips non-matching results. I'm especially weirded out by this choice, because even laravel/collections uses `first(callable)` for this.

File: lib/Doctrine/Common/Collections/Collection.php
Comment (ostrolucky) [inline lib/Doctrine/Common/Collections/Collection.php]: This is not accurate. If Collection is empty, array_reduce will return `null`, no matter what initial value you specify.

Also, no matter what initial value you specify, closure can return some completely different data type than initial value.

This is more like it https://github.com/vimeo/psalm/issues/846

File: lib/Doctrine/Common/Collections/Collection.php
Comment (ostrolucky) [inline lib/Doctrine/Common/Collections/Collection.php:224]: `findOne` returning null when result is not found makes it:
- Inconsistent with what `last()` and `first()` methods return. I think we should first solve this inconsistency before merging this PR
- Impossible to differentiate between missing result and collection containing `null` value

**Review by ostrolucky — CHANGES_REQUESTED** [review]

File: lib/Doctrine/Common/Collections/Collection.php
Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php:224]: I previously return `false` but got a review from @greg0ire https://github.com/doctrine/collections/pull/252#discussion_r475088024

I can return `false` for consistency, if you prefer. You just need to decide.

By the way, `last` and `first` returns false if not found ; but `get` return null if not found.

File: lib/Doctrine/Common/Collections/Collection.php
Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php]: I previously named this `find`. Then @greg0ire asked for `findOne` https://github.com/doctrine/collections/pull/252#discussion_r475087688.

I don't think as weird the term `find`. It seems familiar to me.
This is similar to JS: https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Array/find
Or to doctrine/ORM `findOne` method.

IMHO, the fact that this is the *first* matching result should be something relevant.
I introduced this method because
- You may need to know that the element exists AND you need to use it.
- Using both `filter` and `first` are under-performant for this case.

File: lib/Doctrine/Common/Collections/Collection.php
Comment (ostrolucky) [reply lib/Doctrine/Common/Collections/Collection.php:224]: Yeah I know we need to decide first. I'm fine with returning null, but we should first make existing code consistent so it returns null, otherwise we create more mess by merging this PR. I do have enough experience in software engineering to know better than to believe in "we will clean it up... later"

File: lib/Doctrine/Common/Collections/Collection.php
Comment (ostrolucky) [reply lib/Doctrine/Common/Collections/Collection.php]: JS doesn't already have `first` function though. ORM does have `find` function, but it has completely different signature than this one. In fact, without knowing what these functions do beforehand, I wouldn't be able to tell the difference between `Collection::findOne` and `ObjectRepository::find` based on name, or I would think `ObjectRepository::find` returns more results than `Collection::findOne`, then be confused why one expects ID and another one closure.

@greg0ire ok to switch back to `first`?

Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php]: > @greg0ire ok to switch back to `first`?

`first` already exist, it
```
Sets the internal iterator to the first element in the collection and returns this element.
```

The method doesn't modify the internal iterator, and should not.
I don't think it's a good idea to modify the signature from `first()` to `first(?callable)`.

A specific function seems better.

> ORM does have find function, but it has completely different signature than this one. In fact, without knowing what these functions do beforehand, I wouldn't be able to tell the difference between Collection::findOne and ObjectRepository::find based on name, or I would think ObjectRepository::find returns more results than Collection::findOne, then be confused why one expects ID and another one closure.

I don't know why you compare to `find`, you should compare to `findOne` instead.
Both `Collection::findOne` and `ObjectRepository::findOne` will return 1 value.

File: lib/Doctrine/Common/Collections/Collection.php
Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php:224]: I can see some logic in the current situation.

When we want to access to a value (like a `get`), `null` is return when not found.
When we want to perform an action 'like a `set`), `false` is return when a failure occurs.

`first` sets the index to the first element. If there is no element, it **cannot set**. Failure => `false`.
Same idea for `last`, `current`, `next`.

`key` gets the key of the element at the current iterator position. If there is no element, it **cannot get**. Nothing =>`null`.
Same idea for `get`. It would be the same idea for `findOne`.

The only remaining confusing method would be `remove` then.
It removes the element at a specific `key`, and return `null` if the element was already removed. By following the same logic, it should return `false` instead. (By the way, `removeElement` return `false` for the similar situation).
But [that's not the only issue I have with the `remove` method](https://github.com/doctrine/collections/issues/87#issuecomment-673672873): For consistency, I would prefer `remove` and `removeKey` instead of `removeElement` and `remove`.

File: lib/Doctrine/Common/Collections/Collection.php
Comment (ostrolucky) [reply lib/Doctrine/Common/Collections/Collection.php]: > I don't know why you compare to find, you should compare to findOne instead.

`ObjectRepository::findOne` doesn't exist either. Are you thinking of `findOneBy`?

> The method doesn't modify the internal iterator, and should not.

This is good argument, but TBH I don't see the utility of functions like current(), next() and key() or other ways to play with internal pointer. I never saw a use for those, it was always foreach. We might need to think of removing them. Then nobody would have issues related to internal pointer, as it wouldn't be exposed.

> IMHO, the fact that this is the first matching result should be something relevant.

WDYM? This is why I suggested to use `first`. I don't complain about `One` part, I'm on board with that. Otherwise `findOne` name doesn't tell it returns first matching result. Yes, it's consistent with current implicit Persistence API, doesn't mean we can't create better, more explicit API going forward. Perhaps `findFirst()`?

Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php]: > > I don't know why you compare to find, you should compare to findOne instead.
> 
> `ObjectRepository::findOne` doesn't exist either. Are you thinking of `findOneBy`?

Yes indeed ! I was thinking of `findOneBy`, sorry for my mistake.

> > The method doesn't modify the internal iterator, and should not.
> 
> This is good argument, but TBH I don't see the utility of functions like current(), next() and key() or other ways to play with internal pointer. I never saw a use for those, it was always foreach. We might need to think of removing them. Then nobody would have issues related to internal pointer, as it wouldn't be exposed.

I never played with iterator either.

> > IMHO, the fact that this is the first matching result should be something relevant.
> 
> WDYM?

I was kinda asking myself if the fact that it was the first matching result was something important.

For instance if I had something as 
```
new ArrayCollection(
    ['id' => 1, 'ready' => true],
    ['id' => 2, 'ready' => false],
    ['id' => 3, 'ready' => false],
    ['id' => 4, 'ready' => true],
    ['id' => 5, 'ready' => true],
)
```
And I look for a slot which is ready, I won't care if it's the id `1`, `4` or `5`.
I also don't care if sometimes this configuration gives me `1` and sometimes it gives me `5`. I just want one.

If the fact that it's the `first` result is not relevant, `findOne` seems ok.
And then maybe the doc should be updated to
```
Returns one element of this collection that satisfies the given predicate.
```

If the fact the it's the `first` result is important, `findFirst` could be indeed better.

**Comment by VincentLanglet** [comment]
@ostrolucky Sorry for the delay ; I made the changes you asked :)

File: lib/Doctrine/Common/Collections/Collection.php
Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php]: Php does have a `array_search` method ; maybe `searchValue` is better ?

File: lib/Doctrine/Common/Collections/Collection.php
Comment (VincentLanglet) [reply lib/Doctrine/Common/Collections/Collection.php]: Done

File: lib/Doctrine/Common/Collections/Collection.php
Comment (ostrolucky) [inline lib/Doctrine/Common/Collections/Collection.php]: Parameters are required to specify. Please make sure whole definition works in psalm.dev https://psalm.dev/r/61a6394fe6

**Review by ostrolucky — CHANGES_REQUESTED** [review]

**Comment by ostrolucky** [comment]
Still not it https://psalm.dev/r/b66b6e6a60

**Comment by VincentLanglet** [comment]
> Still not it https://psalm.dev/r/b66b6e6a60

I think your test is wrong since you're passing string to array_reduce. @ostrolucky 

See https://psalm.dev/r/6ca84608e3

**Review by ostrolucky — DISMISSED** [review]

**Comment by VincentLanglet** [comment]
Is there anything missing to be merged ?

**Review by ostrolucky — APPROVED** [review]

**Comment by ostrolucky** [comment]
I don't think so. I am not sure who is responsible for doctrine/collections though.

**Review by greg0ire — APPROVED** [review]

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @VincentLanglet !

**Comment by VincentLanglet** [comment]
> Thanks @VincentLanglet !

You're welcome.

By the way, I saw multiple times that creating a release 2.0 seems impossible since the library will have difficulties to migrate from version 1 and 2. But when I looked at the 2.0 branches, the changes are currently:
- Some method are added to the interface.
- Some return type hint are added.
Which means that if you're compatible with the 2.0 version, you're still compatible with the v1.

What about focusing the feature of the v2 on this point ? (Being compatible with the v2 should allow to be compatible with the v1 easily). For instance: 
- https://github.com/doctrine/collections/issues/231 ; can be easily done since a Closure is a callable, if the method supports the callable, it still supports the closure and work with the v1.
- https://github.com/doctrine/collections/issues/190 ; can not be done since if you return null, you're not compatible with the v1. 

With this strategy, a reachable 2.0 milestone could be created and released ; Wdyt ?

**Comment by ostrolucky** [comment]
If 2.0 is created now, it will mean we have to create 3.0 for new features/BC breaks, so that's unlikely. It's not worth creating 2.0 at this point IMHO, there are not that many changes.

**Comment by VincentLanglet** [comment]
> If 2.0 is created now, it will mean we have to create 3.0 for new features/BC breaks, so that's unlikely. It's not worth creating 2.0 at this point IMHO, there are not that many changes.

I didn't ask for creating 2.0 now. But for a reachable milestone.
Almost 0 issue are flagged with the milestone https://github.com/doctrine/collections/issues?q=is%3Aissue+milestone%3A2.0.0+is%3Aopen and the only one will break the support of 1.x.

The last major version was in 2013, I'm sure they are some things which can be improved and allow librairies to support both `^1 || ^2`.

**Comment by Brammm** [comment]
Derp, I was just about to create a PR for a similar findFirst method. Is there any reason this is going in 2.0 and not 1.x?

**Comment by VincentLanglet** [comment]
> Derp, I was just about to create a PR for a similar findFirst method. Is there any reason this is going in 2.0 and not 1.x?

Adding a method to an interface is a BC-break.

**Comment by Brammm** [comment]
Hadn't considered that. Thanks for explaining.

**Comment by fnagel** [comment]
@VincentLanglet Both methods are not part of the UPGRADE.md file or am I missing something?

**Pull request merged**
